File: backend/app/models/ticket.py
import enum
import uuid
from datetime import datetime, timezone
from typing import Optional
from sqlalchemy import String, Text, DateTime, Enum as SAEnum, ForeignKey, Boolean, ARRAY
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID, JSONB
from app.database import Base


class TicketStatus(str, enum.Enum):
    open        = "open"
    in_progress = "in_progress"
    pending     = "pending"
    resolved    = "resolved"
    closed      = "closed"


class TicketPriority(str, enum.Enum):
    critical = "critical"
    high     = "high"
    medium   = "medium"
    low      = "low"


class Ticket(Base):
    __tablename__ = "tickets"

    id:               Mapped[uuid.UUID]          = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    ticket_no:        Mapped[str]                = mapped_column(String(50), unique=True, nullable=False)
    title:            Mapped[str]                = mapped_column(String(500), nullable=False)
    description:      Mapped[Optional[str]]      = mapped_column(Text, nullable=True)
    customer_name:    Mapped[Optional[str]]      = mapped_column(String(200), nullable=True)
    customer_contact: Mapped[Optional[str]]      = mapped_column(String(200), nullable=True)
    product_name:     Mapped[Optional[str]]      = mapped_column(String(200), nullable=True)
    firmware_version: Mapped[Optional[str]]      = mapped_column(String(100), nullable=True)
    status:           Mapped[TicketStatus]       = mapped_column(SAEnum(TicketStatus,   name="ticket_status"),   nullable=False, default=TicketStatus.open)
    priority:         Mapped[TicketPriority]     = mapped_column(SAEnum(TicketPriority, name="ticket_priority"), nullable=False, default=TicketPriority.medium)
    assignee_id:      Mapped[Optional[uuid.UUID]]= mapped_column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    creator_id:       Mapped[uuid.UUID]          = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    ai_analysis:      Mapped[Optional[str]]      = mapped_column(Text, nullable=True)
    ai_suggestions:   Mapped[Optional[str]]      = mapped_column(Text, nullable=True)
    tags:             Mapped[Optional[list]]      = mapped_column(ARRAY(String), nullable=True)
    attachments:      Mapped[Optional[dict]]      = mapped_column(JSONB, default=list)
    created_at:       Mapped[datetime]           = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at:       Mapped[datetime]           = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    resolved_at:      Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)

    creator:  Mapped["User"]               = relationship("User", foreign_keys=[creator_id])
    assignee: Mapped[Optional["User"]]     = relationship("User", foreign_keys=[assignee_id])
    comments: Mapped[list["TicketComment"]]= relationship("TicketComment", back_populates="ticket", cascade="all, delete-orphan")


class TicketComment(Base):
    __tablename__ = "ticket_comments"

    id:          Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    ticket_id:   Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("tickets.id", ondelete="CASCADE"), nullable=False)
    author_id:   Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    content:     Mapped[str]       = mapped_column(Text, nullable=False)
    is_internal: Mapped[bool]      = mapped_column(Boolean, default=False)
    created_at:  Mapped[datetime]  = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    ticket: Mapped["Ticket"] = relationship("Ticket", back_populates="comments")
    author: Mapped["User"]   = relationship("User", foreign_keys=[author_id])

File: frontend/src/types/ticket.ts
export enum TicketStatus {
  open        = "open",
  in_progress = "in_progress",
  pending     = "pending",
  resolved    = "resolved",
  closed      = "closed",
}

export enum TicketPriority {
  critical = "critical",
  high     = "high",
  medium   = "medium",
  low      = "low",
}

export interface UserBrief {
  id:    string;
  name:  string;
  email: string;
  role:  string;
}

export interface TicketComment {
  id:          string;
  ticket_id:   string;
  content:     string;
  is_internal: boolean;
  author:      UserBrief;
  created_at:  string;
}

export interface Ticket {
  id:               string;
  ticket_no:        string;
  title:            string;
  description?:     string;
  customer_name?:   string;
  customer_contact?:string;
  product_name?:    string;
  firmware_version?:string;
  status:           TicketStatus;
  priority:         TicketPriority;
  assignee?:        UserBrief;
  creator:          UserBrief;
  ai_analysis?:     string;
  ai_suggestions?:  string;
  tags?:            string[];
  comments?:        TicketComment[];
  created_at:       string;
  updated_at:       string;
  resolved_at?:     string;
}

// ── 显示配置 ─────────────────────────────────────────────────

export const TICKET_STATUS_CONFIG: Record<
  TicketStatus,
  { label: string; bgColor: string; dotColor: string; textColor: string }
> = {
  open:        { label: "待处理",  bgColor: "bg-blue-50 dark:bg-blue-950/30",   dotColor: "bg-blue-500",   textColor: "text-blue-600 dark:text-blue-400"   },
  in_progress: { label: "处理中",  bgColor: "bg-amber-50 dark:bg-amber-950/30", dotColor: "bg-amber-500",  textColor: "text-amber-600 dark:text-amber-400"  },
  pending:     { label: "等待客户", bgColor: "bg-yellow-50 dark:bg-yellow-950/30",dotColor: "bg-yellow-500",textColor: "text-yellow-600 dark:text-yellow-400"},
  resolved:    { label: "已解决",  bgColor: "bg-green-50 dark:bg-green-950/30", dotColor: "bg-green-500",  textColor: "text-green-600 dark:text-green-400"  },
  closed:      { label: "已关闭",  bgColor: "bg-slate-100 dark:bg-slate-800/40",dotColor: "bg-slate-400",  textColor: "text-slate-500 dark:text-slate-400"  },
};

export const TICKET_PRIORITY_CONFIG: Record<
  TicketPriority,
  { label: string; color: string }
> = {
  critical: { label: "严重", color: "bg-red-100    text-red-700    dark:bg-red-950/50    dark:text-red-400"    },
  high:     { label: "高",   color: "bg-orange-100 text-orange-700 dark:bg-orange-950/50 dark:text-orange-400" },
  medium:   { label: "中",   color: "bg-yellow-100 text-yellow-700 dark:bg-yellow-950/50 dark:text-yellow-400" },
  low:      { label: "低",   color: "bg-green-100  text-green-700  dark:bg-green-950/50  dark:text-green-400"  },
};
